stdlib/asm/collections/smt.masm

```diff
@@ -0,0 +1,163 @@
+    # [d, K, R, ...]
+    adv.smtget adv_push.1
+
+    # if has value, compute node value; use empty root otherwise
+    # [N, d, K, R, V, ...]
+    adv_push.1
+    if.true
+        # make a domain word for depth
+        # [0, 0, d, 0, K, R, ...]
+        push.0 swap push.0.0
+
+        # load the value
+        # [V, 0, 0, d, 0, K, R, V, ...]
+        padw adv_loadw movdnw.3 dupw.3
+
+        # compute remaining path key
+        # [K', V, 0, 0, d, 0, K, R, V, ...]
+        dupw.2 dup.10 eq.16
+        if.true
+            exec.shr_16
+        else
+            dup.10 eq.32
+            if.true
+                exec.shr_32
+            else
+                dup.10 eq.48
+                if.true
+                    exec.shr_48
+                else
+                    dup.10 eq.64
+                    if.true
+                        # remaining path of depth 64 is 0
+                        drop push.0
+                    else
+                        # adv provider returned an invalid depth
+                        # panic
+                        push.0
+                        assert
+                    end
+                end
+            end
+        end
+
+        # perform the permutation and keep only digest output
+        # [N, d, K, R, V, ...]
+        swapw dup.10 movdn.12 hperm dropw swapw dropw
+    else
+        # [d, K, R, V, ...]
+        padw padw drop movdnw.3 push.0 dropw
+
+        # push empty constant as node
+        # [N, d, K, R, V, ...]
+        dup eq.16
+        if.true
+            push.17483286922353768131
+            push.353378057542380712
+            push.1935183237414585408
+            push.4820339620987989650
+        else
+            dup eq.32
+            if.true
+                push.11677748883385181208
+                push.15891398395707500576
+                push.3790704659934033620
+                push.2126099371106695189
+            else
+                dup eq.48
+                if.true
+                    push.10650694022550988030
+                    push.5634734408638476525
+                    push.9233115969432897632
+                    push.1437907447409278328
+                else
+                    dup eq.64
+                    if.true
+                        padw
+                    else
+                        # adv provider returned an invalid depth
+                        # panic
+                        push.0
+                        assert
+                    end
+                end
+            end
+        end
+    end
+
+    # compute index
+    # [i, d, N, K, R, V, ...]
+    movup.4 dup eq.16
+    if.true
+        dup.5 exec.shr_48
+    else
+        dup eq.32
+        if.true
+            dup.5 exec.shr_32
+        else
+            dup eq.48
+            if.true
+                dup.5 exec.shr_16
+            else
+                # depth already checked; can assume 64
+                dup.5
+            end
+        end
+    end
```

Comment:
As mentioned in the first comment above, I think we can do this without conditional statements. I would create a separate helper procedure which does something like this:
```
[d, K[3], ...] -> [idx, rem, ...]
```
Where $idx$ contains $d$ most significant bits of $K[3]$ and $rem$ contains $64 - d$ least significant bits of $K[3]$.

Reply:
This is no longer the case as we moved the concrete implementations depending on the depth. Should we close this discussion?